EventHandler Error handling

PSteinhaus opened the pull request:
This is a first shot at fixing #800.
Instead of having to write error handling wrappers for each EventHandler function one can now just implement `EventHandler::on_error` to achieve customized error handling.

This still doesn't give users much freedom to handle their own kinds of errors though. Their only option for this, at this point, is to create a `GameError::CustomError(String)`, which is much better than nothing, but not ideal yet.

nobbele commented:
Should the error! and eprintln! calls be kept in the event loop? It's already called in the default impl of on_error so it would give double prints.

IMO error! should be kept in the event loop and removed from default on_error so only eprintln! is in on_error and not the event loop.

PSteinhaus commented:
The only reason I split them up into two calls, one in the event loop and one in `on_error`, is that I thought it would be nice to still know where the error is coming from, which can only be done in the event loop, while one may not want to print the whole error (for some reason, but maybe that's not the case), so I moved that part into `on_error`. My thought was basically "if we want to give the user full control over error handling, then why not".

So you would leave error! in the event loop, but remove it from `on_error`? And if I understand correctly you want to remove eprintln! from `on_error`, so that it's only in the event loop? Could you explain why?

nobbele commented:
You correct about my thoughts on all users being logged but user decides on printing. User does have full control of it if it's split like I said, the logger can be implemented however they want.

As for the origin, I am not sure. Does it even need to be known if it already says it in the log? We could also add another parameter with an enum Origin { Update, Draw }.

PSteinhaus commented:
Hmm, I like that idea!
